### PR TITLE
Land devel branch changes on main

### PR DESCRIPTION
**Pull request by rtamalin**
We have pretty much completed the work for preparing devel to merge to main, so let's get it done!

### CODE REVIEW

**Review by KeithMnemonic — APPROVED** [review]
LGTM, This is a big PR so I just did a rudimentary glance. If we miss something we can quickly follow up with a PR but it is better to get this merged sooner so we can work on main moving forward.

**Comment by rtamalin** [comment]
> LGTM, This is a big PR so I just did a rudimentary glance. If we miss something we can quickly follow up with a PR but it is better to get this merged sooner so we can work on main moving forward.

I did review the changes and confirmed that the next differences between the old devel and the rebased devel were just the workflow, package version and associated test, and spec file related changes from main...

**Review by smarlowucf — APPROVED** [review]
LGTM, there will definitely be more cleanup but it will be easier with things in main :+1:

File: csp_billing_adapter/bill_utils.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+        determined volume dimension tier's name to the specified
+        usage value.
+    """
+    for dimension in metric_dimensions:
+        if 'min' in dimension and usage < dimension['min']:
+            continue
+
+        if 'max' in dimension and usage > dimension['max']:
+            continue
+
```
[the remark this answers is not in the document]
Comment (smarlowucf) [reply csp_billing_adapter/bill_utils.py:157]: I.e. the user should not be messing with config thus there's no need for the application to do config validation at runtime.

[the remark this answers is not in the document]
Comment (rjschwei) [reply csp_billing_adapter/bill_utils.py:157]: There are 2 topics here, having the user messing with the config was not something I thought about but certainly possible in the case where there is a VM.

1.) We are not infallible, as such we could make a mistake and configure `max` for our applications in the last dimension
2.) This is an open source project and anyone can use it. We have no documentation, afaik that states that in the last dimension of volume pricing there cannot be a `max` setting

[the remark this answers is not in the document]
Comment (rjschwei) [reply csp_billing_adapter/bill_utils.py:157]: And it is the responsibility of the application to behave according to the business requirements. So if volume pricing is configured and we cannot find a dimension if there is actual usage it is prudent to at least report a warning.

[the remark this answers is not in the document]
Comment (smarlowucf) [reply csp_billing_adapter/bill_utils.py:157]: 1) Where does the validation stop then? If the application doesn't trust the config file it has to validate everything.
2) Yes, but based on time constraints that probably has to fall into post v1 target.

[the remark this answers is not in the document]
Comment (smarlowucf) [reply csp_billing_adapter/bill_utils.py:157]: > And it is the responsibility of the application to behave according to the business requirements. So if volume pricing is configured and we cannot find a dimension if there is actual usage it is prudent to at least report a warning.

Agree, that is a tangible solution we can implement in the given time table.

[the remark this answers is not in the document]
Comment (rjschwei) [reply csp_billing_adapter/bill_utils.py:157]: > Where does the validation stop then? If the application doesn't trust the config file it has to validate everything.

Handling of incorrect configuration in some way shape or form is part of every application. Every bit of data a user touches is potentially incorrect and the application has to deal with it in a reasonable way. kiwi and cloud-init for example have the validation of the config built into the application and have early exit conditions. That does not imply that the adapter cannot have a separate validation step but it does present the use case that the user creates a config, validates the config then makes a "minor" change, does not re-validate and then the application ends up with an invalid config anyway.

**Pull request merged**